File: chiro_engine/rule_engine.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from .condition_parser import ASTNode, evaluate_condition, parse_condition


@dataclass(frozen=True)
class RuleOutput:
    domain: str
    text: str


@dataclass(frozen=True)
class RuleDefinition:
    id: str
    condition: str
    output: RuleOutput
    # Optionaler Schluessel in meta.weightings (z.B. 'main_lines', 'hand_shape', 'mounts').
    # Fehlt er, wird Gewicht 1.0 verwendet.
    category: Optional[str] = None

    @staticmethod
    def from_dict(raw: dict) -> "RuleDefinition":
        output = raw["output"]
        return RuleDefinition(
            id=raw["id"],
            condition=raw["condition"],
            output=RuleOutput(domain=output["domain"], text=output["text"]),
            category=raw.get("category"),
        )


@dataclass(frozen=True)
class MatchedRule:
    id: str
    domain: str
    text: str
    weight: float


@dataclass(frozen=True)
class _CompiledRule:
    rule: RuleDefinition
    ast: ASTNode


class RuleEngine:
    def __init__(self, rules: list[RuleDefinition], weightings: Optional[dict[str, float]] = None):
        self._weightings = weightings or {}
        self._compiled: list[_CompiledRule] = []
        for rule in rules:
            try:
                ast = parse_condition(rule.condition)
            except ValueError as err:
                raise ValueError(f"Failed to parse condition for rule '{rule.id}': {err}") from err
            self._compiled.append(_CompiledRule(rule=rule, ast=ast))

    def evaluate(self, features: dict) -> list[MatchedRule]:
        """Gibt alle Regeln zurueck, deren Bedingung erfuellt ist, sortiert nach Gewicht (absteigend)."""
        matches: list[MatchedRule] = []
        for compiled in self._compiled:
            if evaluate_condition(compiled.ast, features):
                rule = compiled.rule
                weight = self._weightings.get(rule.category, 1.0) if rule.category else 1.0
                matches.append(
                    MatchedRule(id=rule.id, domain=rule.output.domain, text=rule.output.text, weight=weight)
                )
        matches.sort(key=lambda m: m.weight, reverse=True)
        return matches

    def evaluate_by_domain(self, features: dict) -> dict[str, list[MatchedRule]]:
        """Wie evaluate(), aber nach output.domain gruppiert (z.B. 'career', 'relationships')."""
        grouped: dict[str, list[MatchedRule]] = {}
        for match in self.evaluate(features):
            grouped.setdefault(match.domain, []).append(match)
        return grouped

File: src/ruleEngine.ts
import { ASTNode, evaluateCondition, parseCondition } from './conditionParser';

export interface RuleOutput {
  domain: string;
  text: string;
}

export interface RuleDefinition {
  id: string;
  condition: string;
  /** Optional key into meta.weightings (e.g. 'main_lines', 'hand_shape', 'mounts'). Defaults to weight 1.0. */
  category?: string;
  output: RuleOutput;
}

export interface MatchedRule {
  id: string;
  domain: string;
  text: string;
  weight: number;
}

interface CompiledRule extends RuleDefinition {
  ast: ASTNode;
}

export class RuleEngine {
  private readonly compiled: CompiledRule[];
  private readonly weightings: Record<string, number>;

  constructor(rules: RuleDefinition[], weightings: Record<string, number> = {}) {
    this.weightings = weightings;
    this.compiled = rules.map((rule) => {
      try {
        return { ...rule, ast: parseCondition(rule.condition) };
      } catch (err) {
        throw new Error(`Failed to parse condition for rule '${rule.id}': ${(err as Error).message}`);
      }
    });
  }

  /** Returns all rules whose condition matches, sorted by weight (descending). */
  evaluate(features: Record<string, unknown>): MatchedRule[] {
    const matches: MatchedRule[] = [];
    for (const rule of this.compiled) {
      if (evaluateCondition(rule.ast, features)) {
        const weight = rule.category ? this.weightings[rule.category] ?? 1.0 : 1.0;
        matches.push({ id: rule.id, domain: rule.output.domain, text: rule.output.text, weight });
      }
    }
    return matches.sort((a, b) => b.weight - a.weight);
  }

  /** Same as evaluate(), but grouped by output domain (e.g. 'career', 'relationships'). */
  evaluateByDomain(features: Record<string, unknown>): Record<string, MatchedRule[]> {
    const grouped: Record<string, MatchedRule[]> = {};
    for (const match of this.evaluate(features)) {
      (grouped[match.domain] ??= []).push(match);
    }
    return grouped;
  }
}
